### PR TITLE
add what's new in 7.2

### PR DESCRIPTION
**Pull request by BillWagner**
Create the "What's new in C# 7.2 topic.

### CODE REVIEW

**Comment by BillWagner** [comment]
This is ready for review.

/cc @jcouv

File: docs/csharp/whats-new/csharp-7-2.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+are designed to increase performance by minimize copying value types without
+incurring the memory allocations associated with using reference types. The
+features include:
+
+ - The `in` modifier on parameters to specify that an argument is passed by reference, but not modified by the called method.
+ - The `ref readonly` modifier on method returns indicates that a method returns its value by reference but does not allow writes to that object.
+ - The `readonly struct` declaration indicates that a struct is immutable and should be passed as an `in` parameter to its members methods.
+ - The `ref struct` declaration indicates that a struct type accesses managed memory directly and must always be stack allocated.
+
+You can read more about all these changes in [Using value types with reference semantics](../ref-locals-and-returns.md).
```
Comment (jcouv) [inline docs/csharp/whats-new/csharp-7-2.md]: This link is broken, or maybe not published yet. The URL worries me, as it seems to refer to the C# 7.0 feature ("ref locals and returns").

**Comment by jcouv** [comment]
LGTM Thanks
Tagging @jaredpar @MadsTorgersen

File: docs/csharp/whats-new/csharp-7-2.md (hunk 1 above)
Comment (BillWagner) [reply docs/csharp/whats-new/csharp-7-2.md]: It's not published. It's in #3667 

Good point on the link. I'll update it in the other PR once I have reviews on it.

File: docs/csharp/whats-new/csharp-7-2.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+ms.prod: .net
+ms.devlang: devlang-csharp
+---
+# What's new in C# 7.2
+
+C# 7.2 is another point release that adds a number of useful features.
+One theme for this release is working more efficiently with value types by
+avoiding unnecessary copies or allocations. 
+
+The remaining features are small, nice to have features.
```
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-2.md]: nit: nice to have --> nice-to-have

File: docs/csharp/whats-new/csharp-7-2.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+configuration element to select the compiler language version.
+
+The new language features in this release are:
+
+* [Reference semantics with value types](#reference-semantics-with-value-types)
+  - A combination of syntax improvements that enable working with value types using reference semantics.
+* [Non-trailing named arguments](#non-trailing-named-arguments)
+  - Named arguments can be followed by positional arguments.
+* [Leading underscores in numeric literals](#leading-underscores-in-numeric-literals)
+  - Number literals can now have leading underscores before any printed digits.
```
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-2.md]: Number --> Numeric

File: docs/csharp/whats-new/csharp-7-2.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+* [`private protected` access modifier](#private-protected)
+  - The `private protected` access modifier enables access for derived classes in the same assembly.
+
+## Reference semantics with value types
+
+Language features introduced in 7.2 enable scenarios
+for working with value types while using reference semantics. These features
```
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-2.md]: This first sentence is difficult to parse. Perhaps: 
   enable scenarios for working with value types  --> let you work with value types
These features --> They

File: docs/csharp/whats-new/csharp-7-2.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+are designed to increase performance by minimize copying value types without
```
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-2.md]: minimize --> minimizing

File: docs/csharp/whats-new/csharp-7-2.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+incurring the memory allocations associated with using reference types. The
+features include:
+
+ - The `in` modifier on parameters to specify that an argument is passed by reference, but not modified by the called method.
```
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-2.md]: These four features should be parallel in structure. I personally like the format of the first one (<feature>, to <do something>, So my comments will assume all four will follow that format, but you could also make the `in` modifier conform to the other three.
comma after parameters, and no comma before but

File: docs/csharp/whats-new/csharp-7-2.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+ - The `ref readonly` modifier on method returns indicates that a method returns its value by reference but does not allow writes to that object.
```
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-2.md]: returns indicates --> returns, to indicate

File: docs/csharp/whats-new/csharp-7-2.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+ - The `readonly struct` declaration indicates that a struct is immutable and should be passed as an `in` parameter to its members methods.
```
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-2.md]: declaration indicates --> declaration, to indicate 
members methods --> member methods

File: docs/csharp/whats-new/csharp-7-2.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+ - The `ref struct` declaration indicates that a struct type accesses managed memory directly and must always be stack allocated.
```
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-2.md]: declaration indicates --> declaration, to indicate

File: docs/csharp/whats-new/csharp-7-2.md [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+## Non-trailing named arguments
+
+Method calls may now use named arguments that precede positional arguments when those
+named arguments are in the correct positions. Read more in the updated content for
+[Named and optional arguments](../programming-guide/classes-and-structs/named-and-optional-arguments.md).
+
+## Leading underscores in numeric literals
+
+The first implementation of allowing `_` characters in numeric literals did not allow the
```
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-2.md]: "of allowing `_` characters in numeric literals" is not a good description. I'd change it to "The first implementation of support for digit separators in numeric literals...". Or mention 7.0 -- "The implementation of support for digit separators in C# 7.0 did not allow..."

File: docs/csharp/whats-new/csharp-7-2.md [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+`_` to be the first character of the literal value. Numeric
```
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-2.md]: Not all numeric literals, but only binary and hex, no?

File: docs/csharp/whats-new/csharp-7-2.md [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+For example:
+
+```csharp
+int binaryValue = 0b_0101_0101;
+```
+
+## `private protected`
+
+Finally, a new compound access modifier: `private protected`, indicates that a member may be
```
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-2.md]: if you keep the colon, no comma after `private protected`. If you keep the comma after `private protected`, change the colon to a comma.

File: docs/csharp/whats-new/csharp-7-2.md [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+accessed by derived classes that are declared in the same assembly. Where `protected internal`
```
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-2.md]: Where --> While

File: docs/csharp/whats-new/csharp-7-2.md [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+access allows access by derived classes or classes that are in the same assembly, `private protected`
```
Comment (rpetrusha) [inline docs/csharp/whats-new/csharp-7-2.md]: The first occurrence of "access" here can be deleted.

**Review by rpetrusha — CHANGES_REQUESTED** [review]
A number of comments and suggestions, @BillWagner.

File: docs/csharp/whats-new/csharp-7-2.md (hunk 10 above)
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-2.md]: I agree with @rpetrusha. Also use contractions: did not allow -> didn't allow

File: docs/csharp/whats-new/csharp-7-2.md (hunk 7 above)
Comment (mairaw) [reply docs/csharp/whats-new/csharp-7-2.md]: nit: does not -> doesn't

File: docs/csharp/whats-new/csharp-7-2.md [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+ - The `ref readonly` modifier on method returns indicates that a method returns its value by reference but does not allow writes to that object.
+ - The `readonly struct` declaration indicates that a struct is immutable and should be passed as an `in` parameter to its members methods.
+ - The `ref struct` declaration indicates that a struct type accesses managed memory directly and must always be stack allocated.
+
+You can read more about all these changes in [Using value types with reference semantics](../ref-locals-and-returns.md).
+
+## Non-trailing named arguments
+
+Method calls may now use named arguments that precede positional arguments when those
+named arguments are in the correct positions. Read more in the updated content for
```
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-2.md]: I'd probably stick with the boilerplate sentence here. For more information, see ...

File: docs/csharp/whats-new/csharp-7-2.md [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+```
+
+## `private protected`
+
+Finally, a new compound access modifier: `private protected`, indicates that a member may be
+accessed by derived classes that are declared in the same assembly. Where `protected internal`
+access allows access by derived classes or classes that are in the same assembly, `private protected`
+limits access to derived types declared in the same assembly.
+
+You can read more of the details in the language reference article on [access modifiers](../language-reference/keywords/access-modifiers.md).
```
Comment (mairaw) [inline docs/csharp/whats-new/csharp-7-2.md]: again I'd probably stick with the boilerplate here.

**Review by mairaw — COMMENTED** [review]
LGTM. Some minor comments in addition to the ones already made by @rpetrusha

**Comment by BillWagner** [comment]
Updated based on feedback. @mairaw @rpetrusha Can you take a look?

**Review by rpetrusha — APPROVED** [review]

**Comment by rpetrusha** [comment]
LGTM, @BillWagner. The build is still in progress, but you can merge once the build finishes and it looks good.

**Comment by BillWagner** [comment]
Closing and reopening to force a build (to check a link)

**Pull request merged**